### PR TITLE
Meow.

### PR DESCRIPTION
**Pull request by AKPWebDesign**
Moved entity spawning and related code out to its own class.
Added getInstance static method to Vanish, so we can easily get the plugin instance.
Added ocelot entity spawning effect.
Hopefully followed existing code formatting well enough (let me know if not...)

### CODE REVIEW

File: src/main/java/org/kitteh/vanish/Vanish.java [hunk 1]
```diff
@@ -68,7 +71,12 @@ public static void main(String[] args) {
 
     @Listener
     public void onGameInit(GameInitializationEvent event) {
-        CommandSpec vanishCommandSpec = CommandSpec.builder().permission(PERMISSION_VANISH).executor(new VanishCommand(this)).build();
+    	Vanish.instance = this;
```
Comment (mbax) [inline src/main/java/org/kitteh/vanish/Vanish.java]: tab!

File: src/main/java/org/kitteh/vanish/Vanish.java [hunk 2]
```diff
@@ -38,6 +38,9 @@
     public static final String PERMISSION_VANISH = "vanish.vanish";
 
     public static final String PERMISSION_EFFECTS_BATS = "vanish.effects.bats";
+    public static final String PERMISSION_EFFECTS_CATS = "vanish.effects.cats";
+    
+    public static Vanish instance;
```
Comment (mbax) [inline src/main/java/org/kitteh/vanish/Vanish.java]: public? wut why

File: src/main/java/org/kitteh/vanish/VanishCommand.java [hunk 3]
```diff
@@ -23,42 +23,26 @@
...
-    private final Vanish plugin;
-    private final Cause batCause;
-    private final ParticleEffect effect;
-    private static final int LIFE_TICKS = 3 * 20;
 
-    VanishCommand(Vanish plugin) {
-        this.plugin = plugin;
-        this.batCause = Cause.source(SpawnCause.builder().type(SpawnTypes.PLUGIN).build()).owner(this.plugin).build();
-        this.effect = ParticleEffect.builder().type(ParticleTypes.SMOKE_LARGE).count(1).build();
+	private VanishEntitySpawnEffects entitySpawnEffects;
```
Comment (mbax) [inline src/main/java/org/kitteh/vanish/VanishCommand.java]: tabs all over this area

File: src/main/java/org/kitteh/vanish/VanishEntitySpawnEffects.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+import org.spongepowered.api.world.Location;
+import org.spongepowered.api.world.World;
+
+public class VanishEntitySpawnEffects {
+    private static final int LIFE_TICKS = 3 * 20;
+    private final Cause entityCause;
+    private final ParticleEffect effect;
+    
+    public VanishEntitySpawnEffects() {
+    	this.entityCause = Cause.source(SpawnCause.builder().type(SpawnTypes.PLUGIN).build()).owner(Vanish.getInstance()).build();
```
Comment (mbax) [inline src/main/java/org/kitteh/vanish/VanishEntitySpawnEffects.java]: tabs all over the place

File: src/main/java/org/kitteh/vanish/VanishEntitySpawnEffects.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    	
+    	if (player.hasPermission(Vanish.PERMISSION_EFFECTS_BATS)) {
+    		entityTypes.add(EntityTypes.BAT);
+    	}
+    	
+    	if (player.hasPermission(Vanish.PERMISSION_EFFECTS_CATS)) {
+    		entityTypes.add(EntityTypes.OCELOT);
+    	}
+    	
+    	for (EntityType entityType : entityTypes) {
```
Comment (mbax) [inline src/main/java/org/kitteh/vanish/VanishEntitySpawnEffects.java]: Migrate cleanup outside the for loop!

File: src/main/java/org/kitteh/vanish/Vanish.java (hunk 2 above)
Comment (AKPWebDesign) [inline src/main/java/org/kitteh/vanish/Vanish.java]: because I'm retarded apparently, that's why.

**Comment by mbax** [comment]
Remind me about this later! I completely forgot.

File: src/main/java/org/kitteh/vanish/VanishCommand.java [hunk 6]
```diff
@@ -23,42 +23,28 @@
...
  * Vanish!
  */
 class VanishCommand implements CommandExecutor {
-    private final Vanish plugin;
-    private final Cause batCause;
-    private final ParticleEffect effect;
-    private static final int LIFE_TICKS = 3 * 20;
+
+    private Vanish plugin;
+    private VanishEntitySpawnEffects entitySpawnEffects;
```
Comment (kashike) [inline src/main/java/org/kitteh/vanish/VanishCommand.java]: Can be `final`.

File: src/main/java/org/kitteh/vanish/VanishCommand.java [hunk 7]
```diff
@@ -73,27 +59,9 @@ public CommandResult execute(@Nonnull CommandSource commandSource, @Nonnull Comm
...
-                bats.forEach(bat -> {
-                    if (bat.isLoaded()) {
-                        bat.getWorld().spawnParticles(this.effect, bat.getLocation().getPosition());
-                        bat.remove();
-                    }
-                });
-            }).submit(this.plugin);
-        }
+
+        entitySpawnEffects.applySpawnEffect(player);
```
Comment (kashike) [inline src/main/java/org/kitteh/vanish/VanishCommand.java]: Missing `this.`.

File: src/main/java/org/kitteh/vanish/VanishEntitySpawnEffects.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
+package org.kitteh.vanish;
```
Comment (kashike) [inline src/main/java/org/kitteh/vanish/VanishEntitySpawnEffects.java:24]: Missing license header.

File: src/main/java/org/kitteh/vanish/VanishEntitySpawnEffects.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+import java.util.HashMap;
```
Comment (kashike) [inline src/main/java/org/kitteh/vanish/VanishEntitySpawnEffects.java]: `java.` imports should go below the `org.spongepowered`/etc imports.

File: src/main/java/org/kitteh/vanish/VanishEntitySpawnEffects.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+import org.spongepowered.api.event.cause.entity.spawn.SpawnCause;
+import org.spongepowered.api.event.cause.entity.spawn.SpawnTypes;
+import org.spongepowered.api.world.Location;
+import org.spongepowered.api.world.World;
+
+public class VanishEntitySpawnEffects {
+    private static final int LIFE_TICKS = 3 * 20;
+    private final Cause entityCause;
+    private final ParticleEffect effect;
+    private Vanish plugin;
```
Comment (kashike) [inline src/main/java/org/kitteh/vanish/VanishEntitySpawnEffects.java]: Can be `final`.

File: src/main/java/org/kitteh/vanish/VanishEntitySpawnEffects.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        this.entityCause = Cause.source(SpawnCause.builder().type(SpawnTypes.PLUGIN).build())
+                .owner(this.plugin).build();
+        this.effect = ParticleEffect.builder().type(ParticleTypes.SMOKE_LARGE).count(1).build();
+    }
+
+    public void applySpawnEffect(Player player) {
+        Set<Entity> entities = new HashSet<>();
+        Location<World> location = player.getLocation();
+
+        Map<String, EntityType> permEntityMap = new HashMap<>();
```
Comment (kashike) [inline src/main/java/org/kitteh/vanish/VanishEntitySpawnEffects.java]: This map can be a field in the class.

File: src/main/java/org/kitteh/vanish/VanishEntitySpawnEffects.java [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+public class VanishEntitySpawnEffects {
+    private static final int LIFE_TICKS = 3 * 20;
+    private final Cause entityCause;
+    private final ParticleEffect effect;
+    private final Vanish plugin;
+    private Map<String, EntityType> permEntityMap = new HashMap<>();
```
Comment (kashike) [inline src/main/java/org/kitteh/vanish/VanishEntitySpawnEffects.java]: Can be `final`.

**Comment by AKPWebDesign** [comment]
Should I like, squash this or something?

File: src/main/java/org/kitteh/vanish/VanishEntitySpawnEffects.java [hunk 13]
```diff
@@ -48,6 +50,7 @@
     private final ParticleEffect effect;
     private final Vanish plugin;
     private final Map<String, EntityType> permEntityMap = new HashMap<>();
+    private final Map<Player, Set<Entity>> playerEntityMap = new HashMap<>();
```
Comment (kashike) [inline src/main/java/org/kitteh/vanish/VanishEntitySpawnEffects.java]: You can use a `Multimap`.

File: src/main/java/org/kitteh/vanish/VanishEntitySpawnEffects.java [hunk 14]
```diff
@@ -55,43 +73,60 @@ public VanishEntitySpawnEffects(Vanish plugin) {
...
-        //register to listen to events.
+        // Register to listen to events.
         Sponge.getEventManager().registerListeners(this.plugin, this);
     }
 
+    /**
+     * Spawns entities around a player, based on which permissions they have. Schedules the entities it spawns for later
+     * deletion.
+     *
+     * @param player The player to spawn entities at.
```
Comment (mbax) [inline src/main/java/org/kitteh/vanish/VanishEntitySpawnEffects.java]: lowercase, no period

File: src/main/java/org/kitteh/vanish/VanishEntitySpawnEffects.java [hunk 15]
```diff
@@ -55,43 +73,60 @@ public VanishEntitySpawnEffects(Vanish plugin) {
...
     @Listener
     public void onServerStoppingEvent(GameStoppingServerEvent event) {
+        // Remove all our spawned entities when the server stops.
         this.removeEntities(this.allEntities);
     }
 
+    /**
+     * Removes a collection of entities from the world they're in.
+     *
+     * @param entities The entities to remove.
```
Comment (mbax) [inline src/main/java/org/kitteh/vanish/VanishEntitySpawnEffects.java]: lowercase, no period

**Comment by mbax** [comment]
Thanks!

https://github.com/KittehOrg/Vanish/commit/ebd97f4dd5905d6a22544122e6711aafd3bf82f9

**Pull request closed without merging**